### PR TITLE
Initialization of Sphinx

### PR DESCRIPTION
**Pull request by WenyiQ**
- Some helper tools for document conversion and Sphinx generation
- Converted all Markdown files to reStructured files for better Sphinx supporting
- Changed links into toctree format

### CODE REVIEW

**Review by WenyiQ — COMMENTED** [review]
Not sure if need to keep GitHub as a place to read docs. My understanding is all docs are on Read the Docs.

File: docs/tools/doc_utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import os
+from pathlib import Path
+
+import pypandoc  # type: ignore
+
+base_path = Path(__file__).parent
+filter_path = (base_path / "link_filter.py").resolve()
+
+
+def md_to_rst() -> None:
```
Comment (squirrelsc) [inline docs/tools/doc_utils.py]: This is one time work. this file doesn't need to be committed.

File: docs/tools/link_filter.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from typing import Any
+
+import panflute as pf  # type: ignore
+
+# Thanks to Sergio Correia. Reference:
+# https://stackoverflow.com/questions/40993488/convert-markdown-links-to-html-with-pandoc
```
Comment (squirrelsc) [inline docs/tools/link_filter.py]: Is this called by doc_utils? If so, it doesn't need too.

File: docs/installation_linux.rst [hunk 3]
```diff
@@ -50,9 +50,11 @@ Install Poetry on Linux
 
 Poetry is used to manage Python dependencies of LISA.
 
-:warning: Please enter the root folder of LISA source code to run
-following commands to install poetry, since Poetry manages dependencies
-by the working folder.
+.. warning::
```
Comment (squirrelsc) [inline docs/installation_linux.rst:48]: Is there any tool can help on writting rst in vscode? If there is, document it in development.rst, and all of us can use it.

File: docs/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+sphinx>=4.0
```
Comment (squirrelsc) [inline docs/requirements.txt:1]: Add them to dev section of poetry by below command. The requirements file is not necessary.

`poetry add --dev sphinx sphinx-argparse sphinx-rtd-theme sphinxemoji`

File: docs/make.bat [hunk 5]
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
+
+pushd %~dp0
+
+REM Command file for Sphinx documentation
+
+if "%SPHINXBUILD%" == "" (
```
Comment (squirrelsc) [inline docs/make.bat:7]: Add document about if developer want to test doc changes. It should support both Linux and Windows, and it's better the same commands.

File: pyproject.toml [hunk 6]
```diff
@@ -27,6 +27,8 @@ python-dateutil = "^2.8.1"
 retry = "^0.9.2"
 spurplus = "^2.3.4"
 semver = "^2.13.0"
+pypandoc = "^1.6.3"
```
Comment (squirrelsc) [inline pyproject.toml]: It looks they should be dev dependencies. Is it right?

File: README.rst [hunk 7]
```diff
@@ -36,29 +36,29 @@ Why LISA
...
 ---------
 
--  `Quick start <docs/quick_start.html>`__
--  `Run tests <docs/run_test/run.html>`__
--  `Microsoft tests <docs/run_test/microsoft_tests.html>`__
--  `Write test cases in LISA <docs/write_test/write_case.html>`__
--  `Command line reference <docs/run_test/command_line.html>`__
--  `Runbook reference <docs/run_test/runbook.html>`__
--  `Extend and customize LISA <docs/write_test/extension.html>`__
+-  `Quick start <https://linux-integration-services-automation-lisa.readthedocs.io/en/latest/quick_start.html>`__
```
Comment (squirrelsc) [inline README.rst]: It should avoid to use the absolute links. What's problem here? Is there any other way?

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,75 @@
+Linux Integration Services Automation (LISA)
```
Comment (glima) [inline README.rst:1]: So we'll keep both versions now? Why? What about maintaining the two sets?

File: docs/Makefile [hunk 9]
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
```
Comment (glima) [inline docs/Makefile:1]: Why not have this rule in a top dir Makefile?

File: docs/conf.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+# -- General configuration ---------------------------------------------------
+
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+# ones.
+extensions = [
+    'sphinx.ext.autodoc',
+    'sphinxemoji.sphinxemoji',
+    'sphinx.ext.autosectionlabel',
+]
```
Comment (glima) [inline docs/conf.py:36]: I suggest "sphinx_copybutton", to begin with (I have others in my tree as well, but I'm also learning :))

File: README.rst [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+-  **End-to-end**: ``LISA`` supports platform specific orchestrator to
+   create and delete test environment automatically; it also provides
+   flexibility to preserve environment for troubleshooting if test
+   failed.
+
+Documents
+---------
+
+-  `Quick start <https://linux-integration-services-automation-lisa.readthedocs.io/en/latest/quick_start.html>`__
+-  `Run tests <https://linux-integration-services-automation-lisa.readthedocs.io/en/latest/docs/run_test/run.html>`__
```
Comment (squirrelsc) [inline README.rst]: The url is too long to remember. Can we have a shorter one like `https://lisa.readthedocs.io` or ``https://mslisa.readthedocs.io` if there is conflict?

Comment (glima) [reply README.rst]: Why not MS's https://microsoft.github.io/?

File: docs/conf.py (hunk 10 above)
Comment (WenyiQ) [reply docs/conf.py:36]: Thanks, Gustavo. I will take a look at that.

File: README.rst (hunk 7 above)
Comment (WenyiQ) [reply README.rst]: My thoughts were if we have a specific website for all docs, we do not need bother showing things on GitHub (so we do not prefer this way), since rST files with some Sphinx syntaxes (e.g. [toctree](https://github.com/microsoft/lisa/blob/wenyi/sphinx-init/docs/run_test/run.rst)) are not shown in GitHub.

Comment (WenyiQ) [reply README.rst]: These external links were brought because the links were ending in html and I thought the docs on GitHub must be consistent with the docs on Read the Docs. However, if we have a specific doc branch, we can keep files with correct relative rst links in main and files for the build in doc branch.

File: docs/tools/link_filter.py (hunk 2 above)
Comment (WenyiQ) [reply docs/tools/link_filter.py]: Yes, it's called by doc_utils. But if doc_utils is not needed, this is not needed as well.

File: README.rst (hunk 7 above)
Comment (squirrelsc) [reply README.rst]: ok. For the README, we need the content in both places, and the links go to readthedocs. Other document should have the relative links.

**Comment by squirrelsc** [comment]
There are some conflicts need to be resolved, before merge.

File: docs/installation_linux.rst (hunk 3 above)
Comment (WenyiQ) [reply docs/installation_linux.rst:48]: I'm still learning and I'm currently using a VSCode extension [reStructuredText](https://marketplace.visualstudio.com/items?itemName=lextudio.restructuredtext). I will add that into doc.

File: docs/tools/link_filter.py (hunk 2 above)
Comment (squirrelsc) [reply docs/tools/link_filter.py]: Both of them are not needed, because it's one time work.

File: README.rst (hunk 11 above)
Comment (WenyiQ) [reply README.rst]: @squirrelsc Yes, I can change it (technically, build a new one) to mslisa.
@glima Could you be more specific on that?

File: docs/Makefile (hunk 9 above)
Comment (WenyiQ) [reply docs/Makefile:1]: I have not experimented with this but without this Makefile here it will probably complain during the docs building.

File: docs/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,4 @@
...
+sphinx-argparse
```
Comment (squirrelsc) [inline docs/requirements.txt]: Is this file needed now?

Comment (WenyiQ) [reply docs/requirements.txt]: No it's not. Edited.

Comment (squirrelsc) [reply docs/requirements.txt]: This file is still in latest PR.

Comment (WenyiQ) [reply docs/requirements.txt]: Oh I thought you were asking the specific extension, my bad. This file is needed to tell Read the Docs how to build docs (using Sphinx).

File: docs/conf.py (hunk 10 above)
Comment (WenyiQ) [reply docs/conf.py:36]: Added. Thanks again :-)

File: docs/installation_linux.rst (hunk 3 above)
Comment (WenyiQ) [reply docs/installation_linux.rst:48]: Added in VSCode section in setup.rst.

File: pyproject.toml (hunk 6 above)
Comment (WenyiQ) [reply pyproject.toml]: No longer needed. Deleted.

File: docs/installation_linux.rst (hunk 3 above)
Comment (squirrelsc) [reply docs/installation_linux.rst:48]: BTW, we have files called installation and setup. The names sound similar, and confusing.

Comment (WenyiQ) [reply docs/installation_linux.rst:48]: I personally think they are distinguishing since installation refers to the source code and setup refers to environment etc. I changed the name of setup.rst to dev_setup.rst to make more contrast.

File: docs/troubleshooting.rst [hunk 13]
```diff
@@ -49,6 +49,15 @@ Poetry provides different packages according to the folder, and depends
 on the ``pyproject.toml`` file in the current folder. Make sure to run
 ``poetry`` in the root folder of LISA.
 
+Missing a lot of packages
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, make sure you are currently in the root of your code folder (where you
+cloned LISA source code). Then run ``poetry install`` again to ensure that
+poetry has successfully set up all the dependencies. If you still have problems
+after running this command, refer to `Issue #1401
```
Comment (squirrelsc) [inline docs/troubleshooting.rst]: 1. It doesn't need to refer to the Issue.
2. This FAQ is similar with "Cannot find package after run ``poetry install``". We can update the title to "Missing a lot of packages". It's clearer, because it's part of the error message.
3. The original one doesn't suggest to run `poetry install` again. It should be added.

File: README.rst (hunk 11 above)
Comment (glima) [reply README.rst]: Well, Github has its own doc/webpage hosting system, why use a third party one? Just at what Andy did: https://microsoft.github.io/lisa/

File: docs/Makefile (hunk 9 above)
Comment (glima) [reply docs/Makefile:1]: I have mine at the root dir, no issues at all

File: README.rst (hunk 11 above)
Comment (squirrelsc) [reply README.rst]: It's a good choice. The key difference is the github.io doesn't save pages. The static pages must be submitted to some repo. It's not friendly for hosting multiple version of documents. Either allow a workflow to commit code in CI, or use other repos or branches to save the generated pages. LISA will support versions in future, it means every version release need a separated page cache.

File: README.rst (hunk 8 above)
Comment (squirrelsc) [reply README.rst:1]: The doc website is used to browse. The README.rst is the fallback of document, so it's all links go to doc site. Other files are doc source files.

**Review by squirrelsc — APPROVED** [review]
